README.md
MediaFire for KODI / XBMC
=========================

A MediaFire Video/Music add-on for Kodi / XBMC


changelog.txt
v0.2.0  (2016-02-07)
- fix site changes
- add support for "Following" playback
- implemented: following videos option #3
v0.1.4  (2015-06-03)
- catch captcha prompts
v0.1.3  (2015-04-03)
- add support for premium accounts
v0.1.2  (2014-11-01)
- fix for crashreporting
v0.1.1  (2014-11-01)
- add debug logging support
- add more detail to login errors
- add icons and fanart
v0.1.0  (2014-11-01)
- Initial version


resources/lib/package.py
'''
    Copyright (C) 2014 ddurdle

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


'''

#
#
#
class package:

    ##
    ##
    def __init__(self, file, folder):
        try:
            if file.title != '':
                self.sortTitle = file.title
            else:
                self.sortTitle = folder.title
        except:
            self.sortTitle = folder.title
        self.file = file
        self.folder = folder
        self.mediaurl = None

    def setMediaURL(self,mediaurl):
        self.mediaurl = mediaurl

    def getMediaURL(self):
        return self.mediaurl.url

    def __repr__(self):
        return '{}: {} {}'.format(self.__class__.__name__,
                                  self.sortTitle)

    def __cmp__(self, other):
        if hasattr(other, 'sortTitle'):
            try:
                return self.sortTitle.__cmp__(other.sortTitle)
            except:
                return 0

    def getKey(self):
        return self.sortTitle


resources/lib/mediafire.py
'''
    MediaFire for KODI / XBMC Plugin
    Copyright (C) 2013-2016 ddurdle


    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


'''

import os
import re
import urllib, urllib2
import cookielib
#from resources.lib import authorization
from cloudservice import cloudservice
#from resources.lib import folder
#from resources.lib import file
#from resources.lib import package
#from resources.lib import mediaurl
import unicodedata


import xbmc, xbmcaddon, xbmcgui, xbmcplugin

addon = xbmcaddon.Addon(id='plugin.video.mediafire')
addon_dir = xbmc.translatePath( addon.getAddonInfo('path') )

import os
import sys

sys.path.append(os.path.join( addon_dir, 'resources', 'lib' ) )

import authorization
#import cloudservice
import folder
import file
import package
import mediaurl
import crashreport

#
#
#
class mediafire(cloudservice):


    AUDIO = 1
    VIDEO = 2
    PICTURE = 3

    MEDIA_TYPE_MUSIC = 1
    MEDIA_TYPE_VIDEO = 2
    MEDIA_TYPE_PICTURE = 3

    MEDIA_TYPE_FOLDER = 0

    CACHE_TYPE_MEMORY = 0
    CACHE_TYPE_DISK = 1
    CACHE_TYPE_AJAX = 2

    ##
    # initialize (save addon, instance name, user agent)
    ##
    def __init__(self, PLUGIN_URL, addon, instanceName, user_agent):
        self.PLUGIN_URL = PLUGIN_URL
        self.addon = addon
        self.instanceName = instanceName

        self.crashreport = crashreport.crashreport(self.addon)

        try:
            username = self.addon.getSetting(self.instanceName+'_username')
        except:
            username = ''
        self.authorization = authorization.authorization(username)


        self.cookiejar = cookielib.CookieJar()

        self.user_agent = user_agent

        self.login();



    ##
    # perform login
    ##
    def login(self):

        opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(self.cookiejar), MyHTTPErrorProcessor)
        opener.addheaders = [('User-Agent', self.user_agent),('X-Requested-With' ,'XMLHttpRequest')]

        url = 'https://www.mediafire.com/dynamic/client_login/mediafire.php'

        request = urllib2.Request(url)
        request.add_header('Referer', 'https://www.mediafire.com/templates/login_signup/login_signup.php?dc=loginPath&_dzpbx=4460')

        self.cookiejar.add_cookie_header(request)

        # try login
        try:
            response = opener.open(request)

        except urllib2.URLError, e:
            xbmc.log(self.addon.getAddonInfo('name') + ': ' + str(e), xbmc.LOGERROR)
            return
        response_data = response.read()
        self.cookiejar.extract_cookies(response, request)
        response.close()


        url = 'https://www.mediafire.com/templates/login_signup/login_signup.php?dc=loginPath&_dzpbx=4460'

        request = urllib2.Request(url)
        self.cookiejar.add_cookie_header(request)

        # try login
        try:
            response = opener.open(request)

        except urllib2.URLError, e:
            xbmc.log(self.addon.getAddonInfo('name') + ': ' + str(e), xbmc.LOGERROR)
            return
        response_data = response.read()
        self.cookiejar.extract_cookies(response, request)

        response.close()

        securityValue=''
        for r in re.finditer('name="(security)\" value=\"([^\"]+)\"' ,response_data, re.DOTALL):
            sessionName,securityValue = r.groups()
            #self.authorization.setToken('session_token',securityValue)


        url = 'https://www.mediafire.com/dynamic/client_login/mediafire.php'

        request = urllib2.Request(url)
        request.add_header('Referer', 'https://www.mediafire.com/templates/login_signup/login_signup.php?dc=loginPath&_dzpbx=4460')

        self.cookiejar.add_cookie_header(request)

        # try login
        try:
            response = opener.open(request,'security='+securityValue+'&login_email='+self.authorization.username+'&login_pass='+self.addon.getSetting(self.instanceName+'_password')+'&login_remember=on')

        except urllib2.URLError, e:
            xbmc.log(self.addon.getAddonInfo('name') + ': ' + str(e), xbmc.LOGERROR)
            return
        response_data = response.read()
        response.close()
#        self.crashreport.sendError('test',response_data)

        for cookie in self.cookiejar:
            for r in re.finditer(' ([^\=]+)\=([^\s]+)\s',
                        str(cookie), re.DOTALL):
                cookieType,cookieValue = r.groups()
                if cookieType == 'session':
                    self.authorization.setToken('session_token',cookieValue)


        sessionValue=''
        for r in re.finditer('(parent)\.bqx\(\"([^\"]+)\"' ,response_data, re.DOTALL):
            sessionName,sessionValue = r.groups()
            self.authorization.setToken('session_token',sessionValue)


        return


    ##
    # return the appropriate "headers" for onedrive requests that include 1) user agent, 2) authorization cookie
    #   returns: list containing the header
    ##
    def getHeadersList(self):
        auth = self.authorization.getToken('auth_token')
        session = self.authorization.getToken('auth_session')
        if (auth != '' or session != ''):
            return [('User-Agent', self.user_agent), ('Cookie', session+'; oc_username='+self.authorization.username+'; oc_token='+auth+'; oc_remember_login=1')]
        else:
            return [('User-Agent', self.user_agent )]



    ##
    # return the appropriate "headers" for onedrive requests that include 1) user agent, 2) authorization cookie
    #   returns: URL-encoded header string
    ##
    def getHeadersEncoded(self):
        auth = self.authorization.getToken('auth_token')
        session = self.authorization.getToken('auth_session')

        if (auth != '' or session != ''):
            return urllib.urlencode({ 'User-Agent' : self.user_agent, 'Cookie' : session+'; oc_username='+self.authorization.username+'; oc_token='+auth+'; oc_remember_login=1' })
        else:
            return urllib.urlencode({ 'User-Agent' : self.user_agent })

    ##
    # retrieve a list of videos, using playback type stream
    #   parameters: prompt for video quality (optional), cache type (optional)
    #   returns: list of videos
    ##
    def getMediaList(self, folderName='', cacheType=CACHE_TYPE_MEMORY):

        if folderName == '':
            folderName = 'myfiles'



        opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(self.cookiejar), MyHTTPErrorProcessor)
        opener.addheaders = [('User-Agent', self.user_agent),('X-Requested-With' ,'XMLHttpRequest')]



        sessionValue = self.authorization.getToken('session_token')
        if (sessionValue == ''):
            xbmcgui.Dialog().ok(self.addon.getLocalizedString(30000), self.addon.getLocalizedString(30049), self.addon.getLocalizedString(30050),'sessionValue')
            self.crashreport.sendError('getMediaList:sessionValue','not set')
            xbmc.log(self.addon.getAddonInfo('name') + ': ' + self.addon.getLocalizedString(30050)+ 'sessionValue', xbmc.LOGERROR)
            return

        if folderName == 'FOLLOWING':
            url = 'https://www.mediafire.com/api/1.4/device/get_foreign_resources.php?r=sfrv&session_token='+sessionValue+'&response_format=json'

            request = urllib2.Request(url)
            self.cookiejar.add_cookie_header(request)

            # try login
            try:
                response = opener.open(request)

            except urllib2.URLError, e:
                xbmc.log(self.addon.getAddonInfo('name') + ': ' + str(e), xbmc.LOGERROR)
                return
            response_data = response.read()
            response.close()


            mediaFiles = []
            # parsing page for files
            for r in re.finditer('\"folders\"\:\[\{.*?\}\]' ,response_data, re.DOTALL):
                    entry = r.group()
                    for q in re.finditer('\"name\"\:\"([^\"]+)\"\,.*?\"folderkey\"\:\"([^\"]+)\"' ,entry, re.DOTALL):
                        subfolderName,subfolderID = q.groups()

                        media = package.package(0,folder.folder(subfolderID,subfolderName))
                        mediaFiles.append(media)
        else:
            url = 'https://www.mediafire.com/api/folder/get_content.php?r=mvbn&content_type=folders&filter=all&order_by=name&order_direction=asc&chunk=1&version=1.2&folder_key='+folderName+'&session_token='+sessionValue+'&response_format=json'


            request = urllib2.Request(url)
            self.cookiejar.add_cookie_header(request)

            # try login
            try:
                response = opener.open(request)

            except urllib2.URLError, e:
                xbmc.log(self.addon.getAddonInfo('name') + ': ' + str(e), xbmc.LOGERROR)
                return
            response_data = response.read()
            response.close()


            mediaFiles = []
            # parsing page for files
            for r in re.finditer('\{\"folderkey\"\:.*?\"dropbox_enabled\"\:\"[^\"]+\"\}' ,response_data, re.DOTALL):
                    entry = r.group()
                    for q in re.finditer('\"folderkey\"\:\"([^\"]+)\"\,\"name\"\:\"([^\"]+)\"' ,entry, re.DOTALL):
                        subfolderID,subfolderName = q.groups()

                        media = package.package(0,folder.folder(subfolderID,subfolderName))
                        mediaFiles.append(media)

            url = 'https://www.mediafire.com/api/folder/get_content.php?r=mvbn&content_type=files&filter=all&order_by=name&order_direction=asc&chunk=1&version=1.2&folder_key='+folderName+'&session_token='+sessionValue+'&response_format=json'

            request = urllib2.Request(url)
            self.cookiejar.add_cookie_header(request)

            # try login
            try:
                response = opener.open(request)

            except urllib2.URLError, e:
                xbmc.log(self.addon.getAddonInfo('name') + ': ' + str(e), xbmc.LOGERROR)
                return
            response_data = response.read()
            response.close()


            # parsing page for files
            for r in re.finditer('\{\"quickkey\"\:.*?\"links\"\:\{[^\}]+\}\}' ,response_data, re.DOTALL):
                    entry = r.group()
                    for q in re.finditer('\"quickkey\"\:\"([^\"]+)\"\,.*?\"filename\"\:\"([^\"]+)\".*?\"normal_download\"\:\"([^\"]+)\"' ,entry, re.DOTALL):
                        fileID,fileName,downloadURL = q.groups()
                        downloadURL = re.sub('\\\\', '', downloadURL)

                        media = package.package(file.file(fileID, fileName, fileName, self.VIDEO, '', ''),folder.folder('',''))
                        media.setMediaURL(mediaurl.mediaurl(downloadURL, '','',''))
                        mediaFiles.append(media)


        return mediaFiles


    ##
    # retrieve a playback url
    #   returns: url
    ##
    def getPlaybackCall(self, playbackType, package):

        downloadURL = package.getMediaURL()
        opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(self.cookiejar), MyHTTPErrorProcessor)
        opener.addheaders = [('User-Agent', self.user_agent),('X-Requested-With' ,'XMLHttpRequest')]

        sessionValue = self.authorization.getToken('session_token')
        if (sessionValue == ''):
            xbmcgui.Dialog().ok(self.addon.getLocalizedString(30000), self.addon.getLocalizedString(30049), self.addon.getLocalizedString(30050),'sessionValue')
            self.crashreport.sendError('getPlaybackCall:sessionValue','not set')
            xbmc.log(self.addon.getAddonInfo('name') + ': ' + self.addon.getLocalizedString(30050)+'sessionValue', xbmc.LOGERROR)
            return

        request = urllib2.Request(downloadURL)

        # if action fails, validate login

        try:
            response = opener.open(request)

        except urllib2.URLError, e:
                xbmc.log(self.addon.getAddonInfo('name') + ': ' + str(e), xbmc.LOGERROR)
                return

        response_data = response.read()
        response.close()

        for r in re.finditer('(id)\=\"(form_captcha)\"' ,response_data, re.DOTALL):
            captcha,downloadURL = r.groups()
            xbmcgui.Dialog().ok(addon.getLocalizedString(30000), addon.getLocalizedString(30054),addon.getLocalizedString(30055), addon.getLocalizedString(30056))
            return None


        downloadURL=''
        for r in re.finditer('(kNO) \= \"([^\"]+)\"\;' ,response_data, re.DOTALL):
            urlName,downloadURL = r.groups()

        if downloadURL == '':
            try:
                if response.info().getheader('Location') != '':
                    return response.info().getheader('Location')
            except:
                pass


        if (downloadURL == ''):
            xbmcgui.Dialog().ok(self.addon.getLocalizedString(30000), self.addon.getLocalizedString(30049), self.addon.getLocalizedString(30050), 'downloadURL')
            self.crashreport.sendError('getPlaybackCall:downloadURL',response_data)
            xbmc.log(self.addon.getAddonInfo('name') + ': ' + self.addon.getLocalizedString(30050)+ 'downloadURL', xbmc.LOGERROR)
            return
        return downloadURL


    ##
    # retrieve a media url
    #   returns: url
    ##
    def getMediaCall(self, package):
        return #not implemented

    ##
    # retrieve a directory url
    #   returns: url
    ##
    def getDirectoryCall(self, folder):
        return self.PLUGIN_URL+'?mode=folder&instance='+self.instanceName+'&directory='+folder.id



class MyHTTPErrorProcessor(urllib2.HTTPErrorProcessor):

    def http_response(self, request, response):
        code, msg, hdrs = response.code, response.msg, response.info()

        # only add this line to stop 302 redirection.
        if code == 302: return response
        if code == 303: return response

        if not (200 <= code < 300):
            response = self.parent.error(
                'http', request, response, code, msg, hdrs)
        return response

    https_response = http_response



default.py
'''
    MediaFire for KODI / XBMC Plugin
    Copyright (C) 2013-2016 ddurdle

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''



import xbmc, xbmcgui, xbmcplugin, xbmcaddon


import sys
import urllib
import cgi
import re

# global variables
PLUGIN_NAME = 'mediafire'



#helper methods
def log(msg, err=False):
    if err:
        xbmc.log(addon.getAddonInfo('name') + ': ' + msg, xbmc.LOGERROR)
    else:
        xbmc.log(addon.getAddonInfo('name') + ': ' + msg, xbmc.LOGDEBUG)

def parse_query(query):
    queries = cgi.parse_qs(query)
    q = {}
    for key, value in queries.items():
        q[key] = value[0]
    q['mode'] = q.get('mode', 'main')
    return q

def addMediaFile(service, isQuickLink, playbackType, package):

    listitem = xbmcgui.ListItem(package.file.title, iconImage=package.file.thumbnail,
                                thumbnailImage=package.file.thumbnail)

    if package.file.type == package.file.AUDIO:
        infolabels = decode_dict({ 'title' : package.file.title })
        listitem.setInfo('Music', infolabels)
        playbackURL = '?mode=audio'
    elif package.file.type == package.file.VIDEO:
        infolabels = decode_dict({ 'title' : package.file.title , 'plot' : package.file.plot })
        listitem.setInfo('Video', infolabels)
        playbackURL = '?mode=video'
    else:
        infolabels = decode_dict({ 'title' : package.file.title , 'plot' : package.file.plot })
        listitem.setInfo('Video', infolabels)
        playbackURL = '?mode=video'

    listitem.setProperty('IsPlayable', 'true')
    listitem.setProperty('fanart_image', package.file.fanart)
    cm=[]

    url = package.getMediaURL()
    cleanURL = re.sub('---', '', url)
    cleanURL = re.sub('&', '---', cleanURL)
    url = PLUGIN_URL+'?mode=streamurl&title='+package.file.title+'&url='+cleanURL

#    listitem.addContextMenuItems( commands )
    if cm:
        listitem.addContextMenuItems(cm, False)
    xbmcplugin.addDirectoryItem(plugin_handle, url, listitem,
                                isFolder=False, totalItems=0)

def addDirectory(service, folder):
    listitem = xbmcgui.ListItem(decode(folder.title), iconImage='', thumbnailImage='')
    fanart = addon.getAddonInfo('path') + '/fanart.jpg'

    if folder.id != '':
        cm=[]
#        cm.append(( addon.getLocalizedString(30042), 'XBMC.RunPlugin('+PLUGIN_URL+'?mode=buildstrm&title='+folder.title+'&instanceName='+str(service.instanceName)+'&folderID='+str(folder.id)+')', ))
        listitem.addContextMenuItems(cm, False)
    listitem.setProperty('fanart_image', fanart)
    xbmcplugin.addDirectoryItem(plugin_handle, service.getDirectoryCall(folder), listitem,
                                isFolder=True, totalItems=0)

def addMenu(url,title):
    listitem = xbmcgui.ListItem(decode(title), iconImage='', thumbnailImage='')
    fanart = addon.getAddonInfo('path') + '/fanart.jpg'

    listitem.setProperty('fanart_image', fanart)
    xbmcplugin.addDirectoryItem(plugin_handle, url, listitem,
                                isFolder=True, totalItems=0)


#http://stackoverflow.com/questions/1208916/decoding-html-entities-with-python/1208931#1208931
def _callback(matches):
    id = matches.group(1)
    try:
        return unichr(int(id))
    except:
        return id

def decode(data):
    return re.sub("&#(\d+)(;|(?=\s))", _callback, data).strip()

def decode_dict(data):
    for k, v in data.items():
        if type(v) is str or type(v) is unicode:
            data[k] = decode(v)
    return data


def numberOfAccounts(accountType):

    count = 1
    max_count = int(addon.getSetting(accountType+'_numaccounts'))
    actualCount = 0
    while True:
        try:
            if addon.getSetting(accountType+str(count)+'_username') != '':
                actualCount = actualCount + 1
        except:
            break
        if count == max_count:
            break
        count = count + 1
    return actualCount


#global variables
PLUGIN_URL = sys.argv[0]
plugin_handle = int(sys.argv[1])
plugin_queries = parse_query(sys.argv[2][1:])

addon = xbmcaddon.Addon(id='plugin.video.mediafire')

addon_dir = xbmc.translatePath( addon.getAddonInfo('path') )

import os
sys.path.append(os.path.join( addon_dir, 'resources', 'lib' ) )

import mediafire
import cloudservice
import folder
import file
import package
import mediaurl
import authorization


#from resources.lib import mediafire
#from resources.lib import gPlayer
#from resources.lib import tvWindow
#from resources.lib import cloudservice
#from resources.lib import folder
#from resources.lib import file
#from resources.lib import package
#from resources.lib import mediaurl



#debugging
try:

    remote_debugger = addon.getSetting('remote_debugger')
    remote_debugger_host = addon.getSetting('remote_debugger_host')

    # append pydev remote debugger
    if remote_debugger == 'true':
        # Make pydev debugger works for auto reload.
        # Note pydevd module need to be copied in XBMC\system\python\Lib\pysrc
        import pysrc.pydevd as pydevd
        # stdoutToServer and stderrToServer redirect stdout and stderr to eclipse console
        pydevd.settrace(remote_debugger_host, stdoutToServer=True, stderrToServer=True)
except ImportError:
    log(addon.getLocalizedString(30016), True)
    sys.exit(1)
except :
    pass


# retrieve settings
user_agent = addon.getSetting('user_agent')


mode = plugin_queries['mode']

# make mode case-insensitive
mode = mode.lower()


log('plugin url: ' + PLUGIN_URL)
log('plugin queries: ' + str(plugin_queries))
log('plugin handle: ' + str(plugin_handle))

#if mode == 'main':
#    addMenu(PLUGIN_URL+'?mode=options','<<'+addon.getLocalizedString(30043)+'>>')


#dump a list of videos available to play
if mode == 'main' or mode == 'folder':

    cacheType = int(addon.getSetting('playback_type'))

    folderName=''

    if (mode == 'folder'):
        folderName = plugin_queries['directory']
    else:
        pass

    try:
        isQuickLink = addon.getSetting('playback_type')
        if isQuickLink == 'true':
            isQuickLink = True
        else:
            isQuickLink = False
    except:
        isQuickLink = False

    instanceName = ''
    try:
        instanceName = (plugin_queries['instance']).lower()
    except:
        pass

    numberOfAccounts = numberOfAccounts(PLUGIN_NAME)

    # show list of services
    if numberOfAccounts > 1 and instanceName == '':
        count = 1
        max_count = int(addon.getSetting(PLUGIN_NAME+'_numaccounts'))
        while True:
            instanceName = PLUGIN_NAME+str(count)
            try:
                username = addon.getSetting(instanceName+'_username')
                if username != '':
                    addMenu(PLUGIN_URL+'?mode=main&instance='+instanceName,username)
            except:
                break
            if count == max_count:
                break
            count = count + 1

    else:
        # show index of accounts
        if instanceName == '' and numberOfAccounts == 1:

                count = 1
                max_count = int(addon.getSetting(PLUGIN_NAME+'_numaccounts'))
                loop = True
                while loop:
                    instanceName = PLUGIN_NAME+str(count)
                    try:
                        username = addon.getSetting(instanceName+'_username')
                        if username != '':

                            #let's log in
                            service = mediafire.mediafire(PLUGIN_URL,addon,instanceName, user_agent)
                            loop = False
                    except:
                        break

                    if count == max_count:
                        break
                    count = count + 1

        # no accounts defined
        elif numberOfAccounts == 0:

            #legacy account conversion
            try:
                username = addon.getSetting('username')

                if username != '':
                    addon.setSetting(PLUGIN_NAME+'1_username', username)
                    addon.setSetting(PLUGIN_NAME+'1_password', addon.getSetting('password'))
                    addon.setSetting(PLUGIN_NAME+'1_auth_token', addon.getSetting('auth_token'))
                    addon.setSetting(PLUGIN_NAME+'1_auth_session', addon.getSetting('auth_session'))
                    addon.setSetting('username', '')
                    addon.setSetting('password', '')
                    addon.setSetting('auth_token', '')
                    addon.setSetting('auth_session', '')
                else:
                    xbmcgui.Dialog().ok(addon.getLocalizedString(30000), addon.getLocalizedString(30015))
                    log(addon.getLocalizedString(30015), True)
                    xbmcplugin.endOfDirectory(plugin_handle)
            except :
                    xbmcgui.Dialog().ok(addon.getLocalizedString(30000), addon.getLocalizedString(30015))
                    log(addon.getLocalizedString(30015), True)
                    xbmcplugin.endOfDirectory(plugin_handle)

            #let's log in
            service = mediafire.mediafire(PLUGIN_URL,addon,instanceName, user_agent)


        # show entries of a single account (such as folder)
        elif instanceName != '':

            service = mediafire.mediafire(PLUGIN_URL,addon,instanceName, user_agent)

        try:
            service
        except NameError:
            xbmcgui.Dialog().ok(addon.getLocalizedString(30000), addon.getLocalizedString(30051), addon.getLocalizedString(30052), addon.getLocalizedString(30053))
            log(addon.getLocalizedString(30050)+ 'mediafire-login', True)
            xbmcplugin.endOfDirectory(plugin_handle)

        if folderName == '':
            addMenu(PLUGIN_URL+'?mode=folder&instance='+instanceName+'&directory=FOLLOWING','<<Following>>')

        mediaItems = service.getMediaList(folderName,0)

        if mediaItems:
            for item in sorted(mediaItems):

                try:
                    if item.file == 0:
                        addDirectory(service, item.folder)
                    else:
                        addMediaFile(service, isQuickLink, cacheType, item)
                except:
                        addMediaFile(service, isQuickLink, cacheType, item)

        service.updateAuthorization(addon)


#play a video given its exact-title
elif mode == 'video' or mode == 'audio':

    filename = plugin_queries['filename']
    try:
        directory = plugin_queries['directory']
    except:
        directory = ''

    try:
        title = plugin_queries['title']
    except:
        title = ''

    try:
        cacheType = plugin_queries['playback']
    except:
        try:
            cacheType = int(addon.getSetting('playback_type'))
        except:
            cacheType = 0


    instanceName = ''
    try:
        instanceName = plugin_queries['instance']
    except:
        pass

    # show index of accounts
    if instanceName == '':

                count = 1
                max_count = int(addon.getSetting(PLUGIN_NAME+'_numaccounts'))
                while True:
                    instanceName = PLUGIN_NAME+str(count)
                    try:
                        username = addon.getSetting(instanceName+'_username')
                        if username != '':

                            #let's log in
                            service = mediafire.mediafire(PLUGIN_URL,addon,instanceName, user_agent)

                    except:
                        break

                    if count == max_count:
                        break
                    count = count + 1

    elif instanceName != '':

            service = mediafire.mediafire(PLUGIN_URL,addon,instanceName, user_agent)

    try:
            service
    except NameError:
            xbmcgui.Dialog().ok(addon.getLocalizedString(30000), addon.getLocalizedString(30051), addon.getLocalizedString(30052), addon.getLocalizedString(30053))
            log(addon.getLocalizedString(30050)+ 'mediafire-login', True)
            xbmcplugin.endOfDirectory(plugin_handle)


    mediaFile = file.file(filename, title, '', 0, '','')
    mediaFolder = folder.folder(directory,directory)
    url = service.getPlaybackCall(cacheType,package.package(mediaFile,mediaFolder ))

    item = xbmcgui.ListItem(path=url)
    item.setInfo( type="Video", infoLabels={ "Title": title , "Plot" : title } )
    xbmcplugin.setResolvedUrl(int(sys.argv[1]), True, item)
#play a video given its exact-title
elif mode == 'streamurl':

    url = plugin_queries['url']

    try:
        title = plugin_queries['title']
    except:
        title = ''


    instanceName = ''
    try:
        instanceName = plugin_queries['instance']
    except:
        pass

    # show index of accounts
    if instanceName == '':

                count = 1
                max_count = int(addon.getSetting(PLUGIN_NAME+'_numaccounts'))
                loop = True
                while loop:
                    instanceName = PLUGIN_NAME+str(count)
                    try:
                        username = addon.getSetting(instanceName+'_username')
                        if username != '':

                            #let's log in
                            service = mediafire.mediafire(PLUGIN_URL,addon,instanceName, user_agent)
                            loop = False
                    except:
                        break

                    if count == max_count:
                        break
                    count = count + 1

    elif instanceName != '':

            service = mediafire.mediafire(PLUGIN_URL,addon,instanceName, user_agent)

    try:
            service
    except NameError:
            xbmcgui.Dialog().ok(addon.getLocalizedString(30000), addon.getLocalizedString(30051), addon.getLocalizedString(30052), addon.getLocalizedString(30053))
            log(addon.getLocalizedString(30050)+ 'mediafire-login', True)
            xbmcplugin.endOfDirectory(plugin_handle)


    mediaFile = file.file('', '', '', 0, '','')
    mediaFolder = folder.folder('','')
    media = package.package(mediaFile,mediaFolder)
    cleanURL = re.sub('---', '&', url)
#    cleanURL = re.sub('_', '+', cleanURL)

    media.setMediaURL(mediaurl.mediaurl(cleanURL,'','',''))
    url = service.getPlaybackCall(0,media)

    item = xbmcgui.ListItem(path=url)
    item.setInfo( type="Video", infoLabels={ "Title": title , "Plot" : title } )
#    item.setInfo( type="Music", infoLabels={ "Title": title , "Plot" : title } )

    xbmcplugin.setResolvedUrl(int(sys.argv[1]), True, item)


if mode == 'options' or mode == 'buildstrm' or mode == 'clearauth':
    addMenu(PLUGIN_URL+'?mode=clearauth','<<'+addon.getLocalizedString(30018)+'>>')
    addMenu(PLUGIN_URL+'?mode=buildstrm','<<'+addon.getLocalizedString(30025)+'>>')



xbmcplugin.endOfDirectory(plugin_handle)

